### PR TITLE
[v6r21] Multichoice matching strategy for multiprocessor resources 

### PR DESCRIPTION
**Pull request by atsareg**
This PR adds the possibility to be more flexible in the job matching when using multiprocessor nodes. 
First, for jobs specifying multi-processor requirement (NumberOfProcessors or WholeNode) an extra tag MultiProcessor is added. On the other hand if a worker node presents multi-processor capability, the MultiProcessor tag will be also added automatically in order to satisfy the job requirements. Adding the MultiProcessor tag however will allow the worker node to present RequiredTag = Multiprocessor requirement to match only multi-processor jobs and avoid taking single-processor jobs. Otherwise, this change should not affect the current functionality. This reflects the suggestion made in #3926.
Second, the JobAgent is instrumented to get ComputingElement capability description as a list of several options which represent a prioritized list of what a worker node expects to get. If the CE returns a usual single description, the behavior is the same as before.
Third, the PoolComputingElement is returning optionally a list of descriptions each element differing by the RequiredTag value. If enabled, then there will be, first, an attempt to get a job with a WholeNode requirement if the whole worker node is free, second, attempt to get jobs with the NumberOfProcessors equal to the number of available processors on the node, third, jobs with more than 1 processor requirement and, finally, the description with no extra requirements on the jobs. This will allow to give priority to multiprocessor jobs on the multiprocessor nodes although single-processor jobs are not excluded. Later we can possibly add other strategies as well. 

Note that with the default configuration this PR does not affect the current behavior.

BEGINRELEASENOTES

*WMS
NEW: JobAgent - added possibility to try out several CE descriptions when getting jobs in one cycle
NEW: Matcher - MultiProcessor tag is added to the resource description if appropriate
NEW: JobScheduling - MultiProcessor tag is added to the job description if it specifies multiple processor requirements

*Resources
NEW: PoolComputingElement - getDescription returns a list of descriptions with different requirements to jobs to be matched

ENDRELEASENOTES

### CODE REVIEW

File: WorkloadManagementSystem/Client/Matcher.py [hunk 1]
```diff
@@ -202,6 +202,10 @@ def _processResourceDescription(self, resourceDescription):
         if paramTags:
           resourceDict.setdefault("Tag", []).extend(paramTags)
 
+    # Add 'MultiProcessor' to the list of tags
+    if nProcessors:
```
Comment (fstagni) [inline WorkloadManagementSystem/Client/Matcher.py]: I think it should be `if nProcessors > 1`

File: WorkloadManagementSystem/Agent/JobAgent.py [hunk 2]
```diff
@@ -177,28 +180,38 @@ def execute(self):
...
+        ceDict.update(requirementsDict)
+        self.log.info('Requirements:', requirementsDict)
+
+      self.log.verbose(ceDict)
+      start = time.time()
+      jobRequest = MatcherClient().requestJob(ceDict)
+      matchTime = time.time() - start
+      self.log.info('MatcherTime = %.2f (s)' % (matchTime))
+      if jobRequest['OK']:
+        break
```
Comment (fstagni) [inline WorkloadManagementSystem/Agent/JobAgent.py]: Why?

Comment (atsareg) [reply WorkloadManagementSystem/Agent/JobAgent.py]: If you got a job, you exit the loop on different job request strategies and continue with the usual JobAgent logic to process the received job.

File: Resources/Computing/PoolComputingElement.py [hunk 3]
```diff
@@ -194,6 +194,38 @@ def getCEStatus(self, jobIDList=None):
...
+  def getDescription(self):
+    """ Get CE description as a dictionary
+    """
+    result = super(PoolComputingElement, self).getDescription()
+    if not result['OK']:
+      return result
+    ceDict = result['Value']
+
+    ceDictList = []
+    if self.ceParameters.get('MultiProcessorStrategy'):
```
Comment (fstagni) [inline Resources/Computing/PoolComputingElement.py]: Where is this set?

Comment (atsareg) [reply Resources/Computing/PoolComputingElement.py]: This is in the ComputingElement configuration. Either in PoolComputingElement in general or rather in he local CEDefaults as usually done by a pilot

Comment (fstagni) [reply Resources/Computing/PoolComputingElement.py]: I just grepped the code (and the docs) and could not find it anywhere. Is it a matter of documentation?

Comment (atsareg) [reply Resources/Computing/PoolComputingElement.py]: I will add it to the docs

**Pull request merged**